### ISSUE TITLE
aio: Provide link to working example of rollup and treeshaking for ToH

### ISSUE THREAD

**Issue by dmitrysteblyuk** [angular/angular#17895]
## I'm submitting a ...

<pre><code>
[ ] Regression (behavior that used to work and stopped working in a new release)
[ ] Bug report 
[ ] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
There's a guide on how to add rollup and treeshaking to the tour of heroes: https://angular.io/guide/aot-compiler#tree-shaking. There's no link to actual implementation. Zip-archive of example of ToH doesn't contain rollup config.

## Expected behavior
It would be really nice to see if it's really working.

## What is the motivation / use case for changing the behavior?
If tree-shaking not working as it's stated on the website then this section is really misleading. Otherwise an example would be a great place to start from.

```For example, this demo application doesn't use anything from the @angular/forms library. There is no reason to download forms-related Angular code and tree shaking ensures that you don't.```.

The way I see it, @angular/forms is still in the bundle whole.

**Please tell us about your environment:**
Windows 10

* **Angular version:** 4.2.4

* **Browser:** any
 
* **Language:** Typescript 2.3.3

**Comment by dmitrysteblyuk** [comment]
In truth, I haven't seen real working rollup tree-shaking example for angular. I tested those: 
https://github.com/steveblue/angular2-rollup
https://mlaval.github.io/optimize-angular-app/

The only compiler, I'm aware of, which can tree-shake angular 2 code is Closure.

**Comment by kapunahelewong** [comment]
Hey @petebacondarwin, do you know if this was resolved somewhere in the transition to StackBlitz? (I don't know how to tell.) As far as I know, there isn't an example specific to tree shaking anywhere in the docs, and there's only a brief mention of it in the AOT guide, but I'd be happy to make one. What do you think?

**Comment by petebacondarwin** [comment]
I'll look later today...

**Comment by petebacondarwin** [comment]
@kapunahelewong - AFAICT there is no guide that describes tree shaking any more. The AoT guide only mentions tree-shaking in passing: https://angular.io/guide/aot-compiler#annotationsas.

If you want to put something together that would be fine by me. I am not even sure whether tree-shaking can be done (or is done by default already) in the CLI??  If it is not possible using the CLI then it might be that a stackblitz runnable example is not really possible.

**Comment by kapunahelewong** [comment]
Thank you @petebacondarwin. After some conversation with @filipesilva about it, it sounds like the best way to explain it would be a short paragraph in Deployment about `--build-optimizer` and pointing out the difference in size compared to building without it. 

@dmitrysteblyuk, the reasoning for not creating a specific example here is that the process of treeshaking isn't something you can see take place. Instead, you can only see the evidence of the treeshaking in the size of the build app. 

@petebacondarwin would this be all right?

**Comment by dmitrysteblyuk** [comment]
Yes, only the problem is that I don't see any evidence of working tree-shaking process at all for any bundlers (rollup, webpack etc) except Google Closure (which is error-prone and hard to use with lazy-loaded modules). My guess it's because of angular module system. So I wanted to see somewhere somehow a working example. Angular bundles (especially if you use with material+animations) tend to get pretty big.

**Comment by filipesilva** [comment]
@dmitrysteblyuk do you have the same bundle sizes with and without `--build-optimizer`? There's some tree shaking happening because of optimizations there. There's more tree shaking happening automatically that is not so easy to see though. But generally when you do `ng build --prod` we do all the tree shaking and dead code elimination that we can safely do.

**Comment by dmitrysteblyuk** [comment]
Ok, I made some comparison between Angular Cli and Closure.
Setup is quite simple - create a new angular-cli project and add ```closure.conf```, ```tsconfig.json``` from https://github.com/angular/closure-demo.
Here's the results:

**No FormsModule** (just initial example generated by cli)
Angular Cli - **149 KB** (vendor only, I run ```ng build --prod --no-vendor-chunk=false```)
Closure - **100 KB** (vendor+main)

**FormsModule used a bit** (added only ```<input [(ngModel)]="title">``` and imported FormsModule)
Angular Cli - **190 KB**
Closure - **121 KB**

**FormsModule used extensively** (added multiple input types, select, radio, validators etc)
Angular Cli - **194 KB**
Closure - **135 KB**

So, 

Closure treeshaked **15KB** (it is **40%** of FormsModule - second bundle is that much smaller than the third)
Angular Cli - **4KB** (only **7%** of FormsModule)

Well, results are bit inconclusive, since I didn't separate closure bundle on vendor and main. But on the other hand we can tell that angular cli bundle doesn't get much smaller even when FormsModule is not being used much.

**Comment by filipesilva** [comment]
@dmitrysteblyuk using a vendor chunk greatly reduces the savings of `--build-optimizer` as mentioned in https://github.com/angular/angular-cli/wiki/build#--build-optimizer-and---vendor-chunk. You should compare with just `ng build --prod`.

**Comment by dmitrysteblyuk** [comment]
Ok, now it is 149 - 194 - 207 KB for Angular Cli (13 KB or **22%** treeshaked), which we should take with a grain of salt since app also grows (same is true for closure's 40%). If we remove approximate size of app from the bundle (main.bundle.js was 5 - 6 - 14 KB before I combined it with vendor), we will have 5KB or **10%** of vendor gets treeshaked.

**Comment by petebacondarwin** [comment]
I think it would be very helpful if all this thread was documented in a guide - perhaps deployment.

**Comment by dmitrysteblyuk** [comment]
Angular-cli does treeshaking alright after all. I tried to use only NgForm directive in the template (instead of NgModel) and **~45%** of FormsModule got treeshaked. Not as good as Closure (60%), but at what cost. So I guess an official example is no longer needed seeing as angular cli is well documented anyway and there's an official guide for AoT.
P.S. I opened this issue in the first place, because there was a guide on how to use Rollup, which I couldn't get to work at all.
